index.bs

```diff
@@ -78,4 +251,313 @@ interface StorageBucket {
+
+1. Let |indexedDB| be an {{IDBFactory}} object.
+
+1. Set the [=bottle map=] for |indexedDB| to |bottle map|.
+
+1. Return |indexedDB|.
+
+</div>
+
+<div algorithm>
+
+The user agent MUST consider the [=bucket durability=] when evaluating the <a spec="IndexedDB">durability hint</a> |durability|
+of an {{IDBTransaction}} |transaction|. To <dfn>calculate the effective <a spec="IndexedDB">durability hint</a></dfn> for
+|transaction| with associated |bucket|:
+
+1. If |durability| is not "{{IDBTransactionDurability/default}}", then return |durability|.
+
+1. If |bucket|'s [=bucket durability=] is "{{StorageBucketDurability/strict}}", then return "{{IDBTransactionDurability/strict}}".
+
+1. Return "{{IDBTransactionDurability/relaxed}}".
+
+</div>
+
+<h3 id="storage-bucket-caches">Using CacheStorage</h3>
+
+<div algorithm>
+
+A {{StorageBucket}} has a {{CacheStorage}} object. The <dfn attribute for=StorageBucket>caches</dfn> getter steps are:
+
+1. Let |bucket| be [=this=]'s [=/storage bucket=].
+
+1. Let |bottle map| be the result of [=obtain a local storage bottle map=] with |bucket| and `"cacheStorage"`.
+
+1. Let |cacheStorage| be a {{CacheStorage}} object.
+
+1. Set the <a spec="service-workers">relevant name to cache map</a> for |cacheStorage| to |bottle map|.
+
+1. Return |cacheStorage|.
+
+</div>
+
+<h3 id="storage-bucket-getdirectory">Using an Origin Private File System</h3>
+
+Issue: [[Storage]] needs to define helpers to retrieve the bottle map for a given (non-default) bucket.
+
+Issue: [[FS]] needs to define a helper to retrieve an OPFS given a bottle map.
+
+<div algorithm>
+
+The <dfn method for=StorageBucket>getDirectory()</dfn> steps are:
+
+1. Let |map| be the result of [=obtain a local storage bottle map=] with [=this=]'s [=/storage bucket=] and `"fileSystem"`.
+
+1. Return the result of {{StorageManager/getDirectory}} with |map|.
+
+</div>
+
+<aside class="note">
+
+See [[FS#sandboxed-filesystem]].
+
+</aside>
+
+<h3 id="storage-bucket-clear-site-data">Clear Site Data integration</h3>
+
+Issue: Update [[clear-site-data#header]].
+
+: "<dfn grammar>`storage:bucket-name`</dfn>"
+
+  ::  If the type string starts with "`storage:`" then the remaining characters after the
+      `:` will be taken to refer to a specific [=storage bucket=] in the [=origin=] of a
+      particular response's URL.
+
+Issue: add the steps below to the algorithm in [[clear-site-data#parsing]].
+
+<div algorithm>
+
+To <dfn>parse a Clear-Site-Data header with buckets</dfn>, execute the following steps:
+
+1. For each |type| in |header|, execute the following steps:
+
+    1. If |type| does not [=string/start with=] `"storage:"`, abort these steps.
+
+    1. Let |bucket name| be the [=code-unit-substring-by-positions|code unit substring=] from 8 to |end| of |type|.
+
+    1. If the result of [=validate a bucket name=] with |bucket name| is failure, then abort these steps.
```

Comment:
Do we have a clear use case to clear specific buckets via clear-site-data header?  Do we have folks interested in this willing to help us test it?  This seems more complicated than I expected for a first pass.

Reply:
The code is not particularly complicated, FWIW. Given the examples [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Clear-Site-Data) it seems useful --- e.g. in the case of a user logging out, some user-specific bucket should be cleared, but shared buckets should not be cleared. Having separate buckets for each logged-in user in a multi-login use case is one example use case we often cite for storage buckets.

I don't know of any partners who have asked for it and don't know the history of this idea @ayuishii

Reply:
Looking back at historical discussions, looks like this was a request from @asakusuma from LinkedIn back in 2019. Not sure if they would be active during the Origin Trial for testing though. Haven't heard of other partners that have shown interest otherwise.

Context: https://docs.google.com/document/d/1eBWhY91nUfdT2mys3GaNKX4fKPei79Wk-KlWHYffbAA/edit?usp=sharing

Reply:
Ok.  Seems fair to keep.  Thanks for the context.

Reply:
@wanderview we're not in a position right this moment to do an origin trial, but we are definitely still wanting to be able to use clear-site-data to remotely clear things like just a service worker without collateral damage to other storage items.